File: app/src/main/java/com/harishtk/goldrate/app/net/OkHttpUtil.java
package com.harishtk.goldrate.app.net;

import androidx.annotation.NonNull;

import com.harishtk.goldrate.app.util.ByteUnit;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Objects;

import okhttp3.MediaType;
import okhttp3.ResponseBody;

public final class OkHttpUtil {

    private OkHttpUtil() {}

    public static byte[] readAsBytes(@NonNull InputStream bodyStream, long sizeLimit) throws IOException {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();

        byte[] buffer      = new byte[(int) ByteUnit.KILOBYTES.toBytes(32)];
        int    readLength  = 0;
        int    totalLength = 0;

        while ((readLength = bodyStream.read(buffer)) >= 0) {
            if (totalLength + readLength > sizeLimit) {
                throw new IOException("Exceeded maximum size during read!");
            }

            outputStream.write(buffer, 0, readLength);
            totalLength += readLength;
        }

        return outputStream.toByteArray();
    }
    public static String readAsString(@NonNull ResponseBody body, long sizeLimit) throws IOException {
        if (body.contentLength() > sizeLimit) {
            throw new IOException("Content-Length exceeded maximum size!");
        }

        byte[]    data        = readAsBytes(body.byteStream(), sizeLimit);
        MediaType contentType = body.contentType();
        Charset   charset     = null;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
            charset = contentType != null
                    ? contentType.charset(StandardCharsets.UTF_8) :
                    StandardCharsets.UTF_8;
        } else {
            charset = Charset.forName("UTF-8");
        }

        return new String(data, Objects.requireNonNull(charset));
    }
}

File: app/src/main/java/com/harishtk/goldrate/app/net/OkHttpUtil.kt
package com.harishtk.goldrate.app.net

import com.harishtk.goldrate.app.util.ByteUnit
import okhttp3.ResponseBody
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.util.*

object OkHttpUtil {
    @Throws(IOException::class)
    @JvmStatic
    fun readAsBytes(bodyStream: InputStream, sizeLimit: Long): ByteArray {
        val outputStream = ByteArrayOutputStream()

        val buffer = ByteArray(ByteUnit.KILOBYTES.toBytes(32).toInt())
        var readLength: Int
        var totalLength = 0

        while (bodyStream.read(buffer).also { readLength = it } >= 0) {
            if (totalLength + readLength > sizeLimit) {
                throw IOException("Exceeded maximum size during read!")
            }
            outputStream.write(buffer, 0, readLength)
            totalLength += readLength
        }

        return outputStream.toByteArray()
    }

    @Throws(IOException::class)
    @JvmStatic
    fun readAsString(body: ResponseBody, sizeLimit: Long): String {
        if (body.contentLength() > sizeLimit) {
            throw IOException("Content-Length exceeded maximum size!")
        }

        val data = readAsBytes(body.byteStream(), sizeLimit)
        val contentType = body.contentType()
        val charset = if (contentType != null) {
            contentType.charset(StandardCharsets.UTF_8)
        } else {
            StandardCharsets.UTF_8
        }

        return String(data, Objects.requireNonNull(charset!!))
    }
}